ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/Adversarial.hs

```diff
@@ -0,0 +1,477 @@
+    TestAdversarialMutation recipeH _recipeH' recipeA someRecipeA' mut = testAdversarialMut
+
+    H.HonestRecipe _kcp scg _delta _len = recipeH
+
+    mutatedRecipe = mutateAdversarial recipeA mut
+
+    go catch counter recipeA' testSeedAsSeed = do
+        modifyIORef' counter (+1)
+        let -- TODO is this a low quality random stream? Why is there no @'R.Random' 'QCGen'@ instance?
+            (testSeedA, testSeedAsSeed') = R.split testSeedAsSeed
+
+            schedA = A.uniformAdversarialChain Nothing recipeA' (testSeedA :: QCGen)
+            m      = A.checkAdversarialChain mutatedRecipe schedA
+        -- We discard tests where the first race ends past the acceleration bound
+        -- as these race windows are unconstrained
+        case Exn.runExcept m of
+            Right () -> do
+                let A.UnsafeCheckedAdversarialRecipe { A.carWin } = recipeA'
+                    pretty = case calculateStability scg schedA of
+                        C.SomeWindow Proxy win ->
+                            [H.prettyWindow (C.joinWin carWin win) "no accel"]
+                         <> [show (H.countChainSchema schedA)]
+                writeIORef catch (testSeedA,  H.prettyChainSchema schedA "A" <> pretty)
+                go catch counter recipeA' testSeedAsSeed'
+            Left e   -> case e of
+                A.BadAnchor{} -> error $ "impossible! " <> show e
+                A.BadCount{}  -> error $ "impossible! " <> show e
+                A.BadRace{}   -> pure $ QC.property ()
+
+-----
+
+-- | A nice rendering for failures of 'prop_adversarialChainMutation'
+advMutCounterexample ::
+     TestAdversarialMutation base hon
+  -> A.AdversarialRecipe base hon
+  -> [String]
+  -> QCGen
+  -> String
+advMutCounterexample testAdversarialMut mutatedRecipe schedA' seedA =
+    H.unlines' $ []
+     <> [show seedA]
+     <> [show ch <> " - " <> show arPrefix <> " = " <> show (C.toVar ch - arPrefix) <> "   vs " <> show (kcp, mut)]
+     <> schedH'
+     <> schedA'
+     <> go' (((,) False <$> RI.init kcp' vH) <|> Just (True, RI.initConservative vH))
+  where
+    TestAdversarialMutation recipeH _recipeH' recipeA someRecipeA' mut = testAdversarialMut
+
+    H.HonestRecipe kcp _scg _delta _len = recipeH
+
+    A.AdversarialRecipe { A.arHonest, A.arPrefix } = recipeA
+
+    H.ChainSchema _winH vH = arHonest
+
+    A.AdversarialRecipe { A.arParams = (kcp', _scg', _delta') } = mutatedRecipe
+
+    schedH' = H.prettyChainSchema arHonest "H"
+
+    ch = H.countChainSchema arHonest
+
+    next iter =
+            ((,) False <$> RI.next vH iter)
+        <|>
+            ((,) True <$> RI.nextConservative vH iter)
+
+    go' = \case
+        Nothing -> []
+        Just (cons, iter@(RI.Race (C.SomeWindow Proxy win)))
+          | A.SomeCheckedAdversarialRecipe Proxy recipeA' <- someRecipeA'
+          , A.UnsafeCheckedAdversarialRecipe { A.carWin } <- recipeA'
+          , C.windowStart win < C.windowStart carWin
+          -> go' (next iter)
```

Comment:
d7626941b78800b831f4e7a397ed91183e258c30 changed this to not print the race windows after the first.

But isn't it possible that the "problem" might be in one of those later race windows?

Reply:
I don't think that reasoning works. There are two arguments to make against printing all of the windows.

1. The honest and the adversarial schemas are printed over and over in exactly the same way. The only things that change in the reprints are the indications of where the windows start and end, I think the developer can reasonably infer the windows on her own if the parameters are small. If the parameters are large, the developer won't bother in checking each of the many windows whether printed or not.
2. In mutation tests, the printed examples might not have any problem at all. The example is arbitrarily picked from all examples that pass the schema checks. If the error message could point at an example that has passed the schema check when it shouldn't, printing a verbose output would make more sense.

Reply:
> If the parameters are large, the developer won't bother in checking each of the many windows whether printed or not.

I don't fully agree with this, but I do sympathize with it.

Perhaps the best answer is that the renderer should only print whichever race has the identified violation in it?

Reply:
Ah ha! Nevermind. I didn't realize this code was _only_ used for mutation tests.

Reply:
> Perhaps the best answer is that the renderer should only print whichever race has the identified violation in it?

All race windows are equal in this regard: none of them causes the adversarial schema check to fail.